openedx/core/djangoapps/programs/tests/test_utils.py

```diff
@@ -967,6 +980,144 @@ def test_multiple_published_course_runs(self):
+        """
+        course1 = _create_course(self, self.course_price, course_run_count=2)
+        course2 = _create_course(self, self.course_price, course_run_count=2)
+        CourseEntitlementFactory(user=self.user, course_uuid=course1['uuid'], mode=CourseMode.VERIFIED)
+        CourseEntitlementFactory(user=self.user, course_uuid=course2['uuid'], mode=CourseMode.VERIFIED)
+        program = ProgramFactory(
+            courses=[course1, course2],
+            is_program_eligible_for_one_click_purchase=True,
+            applicable_seat_types=[CourseMode.VERIFIED],
+        )
+        data = ProgramDataExtender(program, self.user).extend()
+        self.assertFalse(data['is_learner_eligible_for_one_click_purchase'])
+        self.assertEqual(data['skus'], [])
+
+    def test_entitlement_product_wrong_mode(self):
+        """
+        Learner should not be eligible for one click purchase if
+        the entitlement product for a course in the program is not in an applicable mode.
+        """
+        course1 = _create_course(self, self.course_price, course_run_count=0)
+        course2 = _create_course(self, self.course_price)
+        course2['entitlements'].append(EntitlementFactory(mode=CourseMode.PROFESSIONAL))
+        program = ProgramFactory(
+            courses=[course1, course2],
+            is_program_eligible_for_one_click_purchase=True,
+            applicable_seat_types=[CourseMode.VERIFIED],
+        )
+        data = ProgramDataExtender(program, self.user).extend()
+        self.assertFalse(data['is_learner_eligible_for_one_click_purchase'])
+        self.assertEqual(data['skus'], [])
+
+    def test_entitlement_product_and_user_enrollment(self):
+        """
+        Learner should be eligible for one click purchase if they hold an enrollment
+        but not an entitlement in a course for which there exists an entitlement product.
+        """
+        course1 = _create_course(self, self.course_price, make_entitlement=True)
+        course2 = _create_course(self, self.course_price)
+        expected_skus = set([course2['course_runs'][0]['seats'][0]['sku']])
+        CourseEnrollmentFactory(user=self.user, course_id=course1['course_runs'][0]['key'], mode=CourseMode.VERIFIED)
+        program = ProgramFactory(
+            courses=[course1, course2],
+            is_program_eligible_for_one_click_purchase=True,
+            applicable_seat_types=[CourseMode.VERIFIED],
+        )
+        data = ProgramDataExtender(program, self.user).extend()
+        self.assertTrue(data['is_learner_eligible_for_one_click_purchase'])
+        self.assertEqual(set(data['skus']), expected_skus)
+
+    def test_no_id_professional_entitlement_product(self):
+        """
+        Learner should be eligible for one click purchase if 'professional' is an applicable seat type
+        a remaining unpurchased course has a 'no-id-professional' entitlement.
```

Comment:
Nit: missing an 'and' between 'seat type' and 'a remaining unpurchased course'

Reply:
Also, according to @mjfrey we should never create 'no-id-professional' entitlements, so maybe this isn't needed? It might be relevant to test for the case where the user has a 'no-id-professional' enrollment in a course for which 'professional' is an applicable seat type, however.